node/core/approval-voting/src/lib.rs

```diff
@@ -106,6 +107,28 @@ enum Mode {
 	Syncing(Box<dyn SyncOracle + Send>),
 }
 
+enum TaskStatus {
+	InProgress,
+	Executed(ValidationResult),
+}
+
+struct ApprovalState {
+	block_hash: Hash,
+	status: TaskStatus,
+}
+
+impl ApprovalState {
+	fn new(block_hash: Hash) -> Self {
+		Self { block_hash, status: TaskStatus::InProgress }
+	}
+
+	fn verified(block_hash: Hash, validation_result: ValidationResult) -> Self {
+		Self { block_hash, status: TaskStatus::Executed(validation_result) }
+	}
+}
+
+type ApprovalsCache = Arc<Mutex<lru::LruCache<CandidateHash, ApprovalState>>>;
```

Comment:
I needed to pass pass the inner LruCache as mutable, so it is insufficient to use an Arc alone. I opted to introduce a std::sync::Mutex, but I'm not super happy with needing to use a Mutex here. Is there a better way?

Reply:
Yes. The mutex is necessary if you want to share state between the approval task and the main subsystem task. If using a mutex, I do recommend using the `futures` mutex as opposed to the parking-lot one.

Still, I'd recommend not using a mutex at all, and instead taking a different architectural approach.

Architecturally, we tend to avoid explicit synchronization where possible and instead follow a message-passing architecture. So what we're really looking for is a flow where the main subsystem learns about the result of the approval task via messages sent from the approval task when it concludes.

For example, the subsystem already receives `IssueApproval` messages from an approval task when approval work is completed. When receiving that, it could be aware of other blocks we've issued assignments for the same candidate under, and issue approvals for each of those blocks.

We'd also want to have the subsystem learn when the approval task wasn't able to complete. From the perspective of the subsystem, we don't really care if that's because the candidate was invalid or unavailable, we just want to drop the associated trackers.

The other thing that we have to be aware of is that futures can be dropped at any `.await` point. That means that we need to have some kind of scope guard within the approval task which notifies the main subsystem when it is dropped. We can use a `oneshot::Sender` as a scope guard because it notifies its `Receiver` when it's dropped. returning to that later...

In fact, the `BackgroundMessage` construct is fairly obsolete, because it was written in order for this communication pattern:

```rust
Approval Task -> Approval Subsystem -> Other subsystems
```

However, this was written before the introduction of the `SubsystemSender` tool. We can refactor the approval tasks to accept an `impl SubsystemSender` as an argument and issue `AllMessages` that way, instead of making this inconvenient jump through the main task. And then we can refactor the approval subsystem task monitor approval tasks by waiting on a `FuturesUnordered` of `oneshot::Receiver<()>` where `Ok(())` indicates the candidate was validated successfully and `Err(Canceled)` indicates it was dropped.